### ISSUE TITLE
No audio when running on OpenSUSE Leap 15.1

### ISSUE THREAD

**Issue by unholyeyebrows** [flathub/net.sf.fuse_emulator#3]
Hi

When running the Fuse Flatpak on OpenSUSE 15.1 there is no audio.

Duplicate via:
sudo flatpak install net.sf.fuse_emulator
**flatpak run net.sf.fuse_emulator**

The Free Unix Spectrum Emulator (Fuse) version 1.5.7.
(c) 1999-2018 Philip Kendall and others; see the file
'AUTHORS' for more details.

For help, please mail <fuse-emulator-devel@lists.sf.net> or use
the forums at <http://sourceforge.net/p/fuse-emulator/discussion/>.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

**Gtk-Message: 07:43:45.980: Failed to load module "canberra-gtk-module"
ALSA lib ../../src/conf.c:3956:(snd_config_update_r) Cannot access file /etc/alsa-pulse.conf
ALSA lib ../../../src/pcm/pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib ../../src/conf.c:3956:(snd_config_update_r) Cannot access file /etc/alsa-pulse.conf
ALSA lib ../../../src/pcm/pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM plughw:0,0
fuse: error: couldn't open sound device 'default' and 'plughw:0,0' check ALSA configuration.**

The alsa errors match this bug:
https://github.com/tiliado/nuvolaruntime/issues/376

The work-around described in the above also works for Fuse here:
**export ALSA_CONFIG_PATH=; flatpak run net.sf.fuse_emulator**

The Free Unix Spectrum Emulator (Fuse) version 1.5.7.
(c) 1999-2018 Philip Kendall and others; see the file
'AUTHORS' for more details.

For help, please mail <fuse-emulator-devel@lists.sf.net> or use
the forums at <http://sourceforge.net/p/fuse-emulator/discussion/>.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

Gtk-Message: 07:45:20.346: Failed to load module "canberra-gtk-module"

Is there any way this can be fixed in the Flatpak so I can avoid having to run this from the command line each time? Thanks, Nick.

**Comment by bertogg** [comment]
Hi, I cannot debug this this week, but the way you describe it it sounds like a general problem with Flatpak itself, and not specifically with Fuse ?

**Comment by unholyeyebrows** [comment]
> Hi, I cannot debug this this week, but the way you describe it it sounds like a general problem with Flatpak itself, and not specifically with Fuse ?

Seem to be the case. Just for comparison, on Ubuntu 18.04 the Flatpak sound works fine.

**Comment by axredneck** [comment]
Arch Linux, similar issue:
```
ALSA lib ../../src/dlmisc.c:170:(snd_dlsym_verify) unable to verify version for symbol _snd_pcm_plug_open
ALSA lib ../../src/dlmisc.c:296:(snd1_dlobj_cache_get) symbol _snd_pcm_plug_open is not defined inside [builtin]
ALSA lib ../../src/dlmisc.c:170:(snd_dlsym_verify) unable to verify version for symbol _snd_pcm_plug_open
ALSA lib ../../src/dlmisc.c:296:(snd1_dlobj_cache_get) symbol _snd_pcm_plug_open is not defined inside [builtin]
fuse: error: couldn't open sound device 'default' and 'plughw:0,0' check ALSA configuration.
```
![2019-08-19-005318_1920x1080_scrot](https://user-images.githubusercontent.com/20617900/63230883-32489080-c21c-11e9-97bc-34a99a4f5d7b.png)
`export ALSA_CONFIG_PATH=""` doesn't help.

**Comment by bertogg** [comment]
> Arch Linux, similar issue:

That looks like a different problem, though. Do other Flatpak apps have sound? Are you using pulseaudio in your machine?

**Comment by axredneck** [comment]
> Do other Flatpak apps have sound?

Yes, they do.
> Are you using pulseaudio in your machine?

Yes, i do.

**Comment by bertogg** [comment]
Do you have a `.asoundrc` file in your home directory ?

**Comment by axredneck** [comment]
> Do you have a .asoundrc file in your home directory ?

Yes, i do:
[my-asoundrc.txt](https://github.com/flathub/net.sf.fuse_emulator/files/3534730/my-asoundrc.txt)

**Comment by bertogg** [comment]
I'm not an expert in Alsa unfortunately, but can you try removing that file (temporarily) and running Fuse without it? Does it make a difference?

If the problem still happens, can you try this and paste the contents of the output file `/tmp/log.txt` ? You will need to install the `org.freedesktop.Sdk` runtime, however.

`flatpak run --devel --command=strace net.sf.fuse_emulator -e open,openat,access /app/bin/fuse 2> /tmp/log.txt
`

**Comment by axredneck** [comment]
Ok, i deleted .asoundrc and now sound works!
What can be wrong with .asoundrc ?

**Comment by bertogg** [comment]
A Flatpak application should not depend on anything that you have on your home folder, so if your `~/.asoundrc` file makes Fuse behave incorrectly then there's a problem in Fuse and needs to be fixed.

My guess (I'm not an Alsa expert) is that your `~/.asoundrc` configuration is probably fine, but it conflicts with the Alsa setup inside the Flatpak sandbox.

I think what I'm going to do is stop making the home directory available to the Fuse app so the `~/.asoundrc` file is invisible. Alternatively, I can ask to modify the Flatpak runtime so the main configuration file `/usr/share/alsa/alsa.conf` does not include `~/.asoundrc`.

While I'm at it I can also unset the `ALSA_CONFIG_PATH` variable, and fix the original problem reported here.

**Comment by axredneck** [comment]
@bertogg 
> stop making the home directory available to the Fuse app

How i then will run my tap/tzx/trd files stored in my home directory?

**Comment by bertogg** [comment]
> How i then will run my tap/tzx/trd files stored in my home directory?

Two ways to do that:

1. You can make the directories where you have your Spectrum data available to Fuse:

   `$ flatpak run --filesystem=~/spectrum net.sf.fuse_emulator`

   You can make that permanent using flatpak-override (`man flatpak-override` for more details):

   `$ flatpak override --filesystem=~/spectrum net.sf.fuse_emulator`

   (I have seen other Flatpak apps sharing `~/Documents` by default. I expect that many Fuse users have their Spectrum files somewhere else, that's why I decided to share the full home directory instead)

2. Flatpak provides a [FileChoose portal](https://github.com/flatpak/flatpak/wiki/Portals), and it should be possible to make Fuse use it.

   With this, when you go to "File -> Open" or "File -> Save Snapshot" you would get a file selector with complete access to your system, and Fuse would be able to open the files that you select with it even if the directory where they are is not visible from the sandbox.

   This is the "proper" solution, although it would not work with files selected from the command line (only files manually selected by the user). Also, it requires changes to Fuse, and I would need to test them first.

Still, my impression is that the ALSA library should not be reading `~/.asoundrc` in the first place, so I'll try to have that clarified before I proceed with any of these two solutions.

In your case, if you have all your files in e.g. `~/spectrum` you can fix your problem with something like:

`$ flatpak override --nofilesystem=home --filesystem=~/spectrum net.sf.fuse_emulator`

This would hide all the files in your home directory but would still give Fuse access to your Spectrum files.

**Comment by bertogg** [comment]
I just fixed the `ALSA_CONFIG_PATH` problem, update Fuse and it should work now.

I'm closing this bug. About the `.asoundrc` problem, I'm trying to have it fixed in the Flatpak runtime (see [here](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/857) for details).

Thanks!

**Comment by unholyeyebrows** [comment]
Many thanks - I confirm this now works for me in OpenSUSE Leap.

Nick

**Comment by bertogg** [comment]
I forgot to mention it, but the `.asoundrc` problem has already fixed for a while in the 19.08 Freedesktop runtime.
